File: leetcode-submissions/2779_Maximum_Beauty_of_an_Array_After_Applying_Operation.ts
function maximumBeauty(nums: number[], k: number): number {
    const intervals:{start:number ;end:number }[]=[];
    for(let i:number=0;i<nums.length;i++){
        intervals.push({start:nums[i]-k,end:nums[i]+k});

    }
    //after this we need to sort this in ascending order 
    intervals.sort((a,b)=>a.start-b.start);
    const queue:number[]=[];
    let maxBeauty:number =1;
    for(let i:number =0;i<intervals.length;i++){
        while(queue.length>0 && queue[0]<intervals[i].start){
            queue.shift();

        }
        queue.push(intervals[i].end);
        maxBeauty=Math.max(maxBeauty,queue.length);


    }
    return maxBeauty;

};

File: leetcode-submissions/2779_Maximum_Beauty_of_an_Array_After_Applying_Operation.cpp
class Solution {
public:
    int maximumBeauty(vector<int>& nums, int k) {
        // we need to calculate the particular range values 
        vector<pair<int,int>>intervals;
        for(int i=0;i<nums.size();i++){
            intervals.push_back({nums[i]-k,nums[i]+k});
        }
        //after this we need to find out the maximum overlapping intervals 
        deque<pair<int,int>>dq;
        sort(intervals.begin(),intervals.end());
        int result =1;
        
        // we have maximun one overlapping intervals 
        for(int i=0;i<intervals.size();i++){
            while (not dq.empty() and dq.front().second<intervals[i].first){
                dq.pop_front();

            }
            dq.push_back(intervals[i]);
            result=max(result ,(int)dq.size());
        }
        return result ;
    }
};